File: Graph/03_dfs.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // Function to return Depth First Traversal of given graph.
    void dfsT(int node, vector<vector<int>>& adjL, vector<int>& visited,vector<int>& ans){
        visited[node]=1;
        ans.push_back(node);
        for(int adjN:adjL[node]){
            if(!visited[adjN])
              dfsT(adjN,adjL,visited,ans);
        }
    }
    vector<int> dfs(vector<vector<int>>& adj) {
        // Code here
        int n= adj.size();
        vector<int> visited(n,0);
        vector<int> ans;
        dfsT(0,adj,visited,ans);
        return ans;
    }
};

File: Graph/03_dfs.java
class Solution {
    // Function to return a list containing the DFS traversal of the graph.
    
    private void dfsT(int node, ArrayList<ArrayList<Integer>> adj, ArrayList<Integer> ans, ArrayList<Integer> visited){
        visited.set(node,1);
        ans.add(node);
        for(Integer adjN:adj.get(node)){
            if(visited.get(adjN)==-1)
              dfsT(adjN,adj,ans,visited);
        }
    }
    public ArrayList<Integer> dfs(ArrayList<ArrayList<Integer>> adj) {
        // Code here
        ArrayList<Integer> ans=new ArrayList<>();
        int n=adj.size();
        ArrayList<Integer> visited=new ArrayList<>(Collections.nCopies(n,-1));
        
        dfsT(0,adj,ans,visited);
        return ans;
        
    }
}